cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,817 @@
+        ( 1 << cell_bits_per_tile_dim ) - 1;
+    //! Number of bits (total) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile =
+        cell_bits_per_tile_dim + cell_bits_per_tile_dim +
+        cell_bits_per_tile_dim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile =
+        cell_num_per_tile_dim * cell_num_per_tile_dim * cell_num_per_tile_dim;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+
+    /*!
+      \brief Constructor
+      \param size The size of the block (MPI rank) (Unit: cell)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capcity
+    */
+    SparseIndexSpace( const std::array<int, N> size,
+                      const unsigned int capacity, const float rehash_factor )
+        : _block_id_space( size[0] >> cell_bits_per_tile_dim,
+                           size[1] >> cell_bits_per_tile_dim,
+                           size[2] >> cell_bits_per_tile_dim,
+                           1 << bitCount( capacity ), rehash_factor )
+    {
+        std::fill( _min.data(), _min.data() + rank, 0 );
+        std::copy( size.begin(), size.end(), _max.data() );
+    }
+
+    /*!
+      \brief Insert a cell (given a cell ijk, insert the tile where the cell
+      reside in to hash table) \param cell_i cell id in dim-x \param cell_j cell
+      id in dim-y \param cell_k cell id in dim-z
+    */
+    KOKKOS_INLINE_FUNCTION
+    void insertCell( int cell_i, int cell_j, int cell_k ) const
+    {
+        insertTile( cell_i >> cell_bits_per_tile_dim,
+                    cell_j >> cell_bits_per_tile_dim,
+                    cell_k >> cell_bits_per_tile_dim );
+    }
+
+    /*!
+      \brief Insert a tile (to hash table)
+      \param tile_i tile id in dim-x
+      \param tile_j tile id in dim-y
+      \param tile_k tile id in dim-z
+    */
+    KOKKOS_INLINE_FUNCTION
+    void insertTile( int tile_i, int tile_j, int tile_k ) const
+    {
+        _block_id_space.insert( tile_i, tile_j, tile_k );
+    }
+
+    /*!
+      \brief Query the 1D cell key from the 3D cell ijk
+      \param cell_i cell id in dim-x
+      \param cell_j cell id in dim-y
+      \param cell_k cell id in dim-z
+    */
+    KOKKOS_INLINE_FUNCTION
+    value_type queryCell( int cell_i, int cell_j, int cell_k ) const
+    {
+        // query the tile No.
+        auto tile_id = _block_id_space.find( cell_i >> cell_bits_per_tile_dim,
+                                             cell_j >> cell_bits_per_tile_dim,
+                                             cell_k >> cell_bits_per_tile_dim );
+        auto cell_id = _tile_id_space.coordToOffset(
```

Comment:
So for now we are ordering the tiles by whatever the `UnorderedMap` does for us while the cell ordering in a tile is either `Morton` or `Naive`?

Reply:
For now, we didn't sort by hand. Not sure if `UnorderedMap` will handle the ordering with the hashed key in their application.